### PR TITLE
Prevent scrolling up when opening a dialog

### LINKED ISSUE #16873: Editor: link modal & media modal keep scrolling up to the top
When the media modal opens, the page gets scrolled up to the top.

### To Reproduce

1. Open the Editor
2. Have a long post, so it's scrolled down
3. Insert an image, or do any operation opening the media modal while scrolled down (edit).

The page will scroll up as soon as the media modal opens.

I ranked this "High priority" as it's a very frustrating disruption for anyone writing a post with images.

### LINKED ISSUE #27226: Editor: Adding a link using toolbar button causes editor to jump to the top
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/post from a Windows machine using Windows 7 or 10.
2. Add content to a new post. Make sure it's enough content so the first line scrolls off the top of the screen.
3. Towards the end of the post, highlight some text
4. Click the hyperlink button in the toolbar and add a link.

#### What I expected
 
For the link to be added, but to still have the editor focused on the text I'm currently editing. (this is what happens when a link is added using Mac)

#### What happened instead

The editor jumped to the top of the post, and I had to scroll down to get back to the anchor text where I'd added the link.

#### Browser / OS version

I have tested and replicated this on Windows 10 Professional on latest versions of Edge and Chrome.
User reported in Chrome 68.0.3440.106 on Windows 10 (64-bit) and another user in Windows 7 (64-bit) on Opera.

#### Screenshot / Video

User provided a video recording of the issue available here: http://recordit.co/emROVNPLuq. 

#### Context / Source

#user-report
Users reported in 1418879-zen and 1408950-zen

Bug sounds exactly like what was reported here: #16741 and fixed in #16873 but now the issue is only on Windows.

### LINKED ISSUE #29821: Gutenlypso: Media Modal scrolls to top on insert
#### Steps to reproduce
1. Starting at URL: `/block-editor/post`
2. Insert enough content to scroll down.
3. At the bottom of the post create an `/image` block and open the Media Library.
4. Insert an image.

#### What I expected

To remain in the same scroll position I was before opening the Media Library.

#### What happened instead

It scrolled back to the top of the post.

#### Context / Source

Maybe related to #16873?

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

Add a `ReactModal__Html--open` class to the HTML document that unsets [the default `overflow-y: scroll;` CSS declaration](https://github.com/Automattic/wp-calypso/blob/ee1f62933b55ba9de43c4a6267bfb835af478056/assets/stylesheets/shared/_reset.scss#L70) for solving a conflict between that declaration and [the `ReactModal__Body--open` class](https://github.com/Automattic/wp-calypso/blob/0345ff083bb0427b49c5101da4971e7efdfc1dbb/client/components/dialog/style.scss#L114-L116) which is defining a `overflow: hidden;` for the body.

When we scroll in Calypso, we're actually scrolling the body element, so when we disable the scrolling with `overflow: hidden;`, its ancestor, the HTML document, becomes scrollable which by default is scrolled to the top since we have never scrolled that element. Note that although the HTML document is scrollable now, we cannot scroll down when the Dialog is open because the body height is limited to the 100% of the window height.

With the `ReactModal__Html--open`  we avoid this behavior since the HTML content is now visible and not scrollable. This indeed has side effect: the scrollbar is lost so there is a reflow caused by the new window size available. But I don't think is a big deal, since it's almost inappreciable due to the fact that most of the UI changes because of the dialog.

#### Testing instructions

* Create a new post with the Classic editor.
* Add enough content to scroll down.
* Insert an image at the end of the post.
* Check that you remain in the same scroll position you were before opening the Media Library.
* Add a link towards the end of the post (by highlighting some text and clicking on the hyperlink button).
* Make sure you're still in the same scroll position.
* Switch to Gutenberg.
* Insert and image using the Media Library.
* Check again that the page is not scrolled.

Fixes #16873.
Fixes #27226.
Fixes #29821.

### CODE REVIEW

**Comment by Copons** [comment]
`overflow-y: scroll` forces displaying the scrollbar even if the content isn’t long enough to scroll.

The purpose is to prevent the reflow caused by the content obtaining or losing the scrollbar, which happens, for example, when the content length changes, or when you play around with modals.

I can confirm that the change in this PR fixes the modal issue, but I honestly don't know how to effectively test for regressions. 🤔

**Comment by mmtr** [comment]
Ah, you're right @Copons! 

Didn't have that in consideration since I'm using the "auto" scrollbars in macOS, so they appear floating over the content and `overflow: scroll` doesn't affect the size available for the content. 

But indeed that reflow is there if I change the setting to always display the scrollbar (and I guess that this is happening always in Windows).

I'd say that was the single purpose of `overflow-y: scroll` and it's causing a regression.

I guess we could solve it by doing what [Bootstrap does](https://github.com/twbs/bootstrap/blob/bec34acbbe10eca0e08edfb178573d9d02ffb9fe/js/src/modal.js#L437-L471): apply a padding when the modal is opened, so the size available doesn't change. Although this will make the code more complex and I honestly don't like that very much.

The alternative is to do what @jameskoster suggested on #16873, remove the `overflow: hidden` style defined by `ReactModal__Body--open`, but that will make scrollable the content behind the modal.

**Comment by Copons** [comment]
We could also keep track of the scroll position before opening the modal, and make sure to scroll back to it after closing it. 🤔

**Comment by gwwar** [comment]
I think it's probably much less risky for us to adopt a more targeted fix. I don't have any preferences for the approach but will note that JS fixes tend to a bit more finicky and prone to breakage.

**Comment by mmtr** [comment]
I included back the original `overflow-y: scroll` for the HTML document in d4e248e and created a new `ReactModal__Html--open` which unsets that CSS and is only applied when opening a dialog, so the change should be less risky now.

This indeed causes the effect described by @Copons: the scrollbar is gone when the modal is open so there is reflow caused by the new size available in the window, but I don't think this is a problem since the change is almost invisible due to the fact that most of the UI changes because the Dialog is covering most of the previous content.

File: client/components/dialog/dialog-base.jsx [hunk 1]
```diff
@@ -80,7 +77,7 @@ class DialogBase extends Component {
 		}
 
 		return (
-			<div className={ buttonsClassName } ref="actionButtons">
```
Comment (gwwar) [inline client/components/dialog/dialog-base.jsx]: Have we searched to see if there any usages of the actionsButtons or content ref?

Comment (mmtr) [reply client/components/dialog/dialog-base.jsx]: Yup. eslint was complaining about it (string literals are deprecated) so I removed them after confirming they are not used.

**Comment by jameskoster** [comment]
Did y'all see this? https://github.com/Automattic/wp-calypso/pull/28737

Seemed like an elegant fix to me.

**Comment by mmtr** [comment]
Sorry, I wasn't aware about that PR. 

Although it's the same approach I took at the beginning (a3179843bd75d340813c8a26019a6b558f9390c6) so, as @Copons noted (https://github.com/Automattic/wp-calypso/pull/29832#issuecomment-450737171), that didn't prevent the reflow caused by the content obtaining or losing the scrollbar. 

Which is why I decided to implement a more targeted fix, in order to lose the scrollbar only when the dialog is open.

**Pull request merged**